### DISCUSSION TITLE
Change definition of `MemberExpr` and `MetaProp`

### DISCUSSION THREAD

**Discussion by Austaras** [swc-project/swc#3153] [General]
Well yes it's me again.

![图片](https://user-images.githubusercontent.com/15013925/147737669-067ea7a1-0ced-4a59-9485-4edaaaec187d.png)

So once again spec allows a better definition fo AST leveraging rust  `enum`, here is my proposed API
```rs
struct MemberExpr {
    pub span: Span,
    pub obj: Box<Expr>,
    pub prop: MemberProp,
}

enum MemberProp {
    Ident(Ident),
    PrivateName(PrivateName),
    Computed(ComputedPropName),
}

// add
struct SuperPropExpr {
    span: Span,
    prop: SuperProp
}

pub enum SuperProp {
    Ident(Ident),
    Computed(ComputedPropName),
}

struct MetaProp {
    span: Span,
    kind: MetaPropKind
}

enum MetaPropKind {
    NewTarget,
    ImportMeta
    // maybe add a no_exhaustive mark
}
```

However, this will likely be a great lot of works so I'm not sure if it worths it

------
New added: since we get rid of `ExprOrSuper` with `MemberExpr`, we could get better defintion for callee:
![图片](https://user-images.githubusercontent.com/15013925/147858944-cce638ce-d40a-4c53-a8dc-5f1a42e8b965.png)
```rs
enum Call {
  Super, Import,
  Expr(Box<Expr>)
}
``

**Comment by kdy1** [comment]
I think it's a quite good change. `MemberProp.obj` being `Expr` was always a headache because of `Expr::ident`.

**Reply by Austaras** [reply]
Is it right that parse result of `abc.foo = 123` is `PatOrExpr::Pat(Box<Pat::Expr(Expr)>)`?

**Reply by kdy1** [reply]
Honestly I'm not sure about `PatOrExpr`. I think `Pat` itself is enough, but it's lots of work.

**Reply by Austaras** [reply]
Seems to be caused by [this line](https://github.com/swc-project/swc/blob/main/crates/swc_ecma_parser/src/parser/expr.rs#L152), what the purpose of it?

**Reply by kdy1** [reply]
Not all expressions are pattern, and some syntax is valid only for expressions.

e.g. `foo()` should **not** be `Pat::Expr()`. (That's why `PatOrExpr` exists)
At the same time, `{ a = b}` is valid only for pattern, but we need `AssignProp` to be a variant of `Prop` because at the time of parsing we can't know if it's a pattern or an expression.
